src/backend/core/api/permissions.py

```diff
@@ -127,10 +129,74 @@ def has_object_permission(self, request, view, obj):
         ) and deleted_at < get_trashbin_cutoff():
             raise Http404
 
-        # Compute permission first to ensure the "user_roles" attribute is set
-        has_permission = super().has_object_permission(request, view, obj)
+        abilities = obj.get_abilities(request.user)
+        action = view.action
+        try:
+            action = ACTION_FOR_METHOD_TO_PERMISSION[view.action][request.method]
+        except KeyError:
+            pass
+
+        has_permission = abilities.get(action, False)
 
         if obj.ancestors_deleted_at and not RoleChoices.OWNER in obj.user_roles:
             raise Http404
 
         return has_permission
+
+
+class DocumentAccessPermission(IsAuthenticated):
+    """Permission class for document access objects."""
+
+    def has_permission(self, request, view):
+        """check create permission for accesses in documents tree."""
+        if super().has_permission(request, view) is False:
+            return False
+
+        if view.action == "create":
+            role = view.document.get_role(request.user)
+            if role not in choices.PRIVILEGED_ROLES:
+                raise exceptions.PermissionDenied(
+                    "You are not allowed to manage accesses for this resource."
+                )
+
+        return True
+
+    def has_object_permission(self, request, view, obj):
+        """Check permission for a given object."""
+        abilities = obj.get_abilities(request.user)
+
+        requested_role = request.data.get("role")
+        if requested_role and requested_role not in abilities.get("set_role_to", []):
+            return False
+
+        action = view.action
+        return abilities.get(action, False)
+
+
+class TemplateAccessPermission(IsAuthenticated):
+    """Permission class for template access objects."""
+
+    def has_permission(self, request, view):
+        """check create permission for template accesses."""
+        if super().has_permission(request, view) is False:
+            return False
+
+        if view.action == "create":
+            roles = view.template.get_roles(request.user)
+            if not set(roles).intersection(set(choices.PRIVILEGED_ROLES)):
```

Comment:
This is the only difference between `TemplateAccessPermission` and `DocumentAccessPermission`. Is there a way to factorize it ?

Reply:
Good catch! Consider it done.